podkrepi-bg/frontend
# Campaign Applications (alpha)

## Editing and file upload and admin section for campaign applications - finalizes first alpha version of the whole process

### feat: file upload and summary
  - handle file upload one by one to allow for handling errors individually (otherwise the server just returns response failed for any one file that fails and no information about the successful uploads)
  - add summary where the uploaded and failed (if any) files are top and center
  - add a short  explanation of what to do about failed files
- feat: validation of create campaign application
- fix: send campaignTypeId + editing + visuals

### send campaignTypeId instead of the category (which is a property of the campaign type and is not enough to uniquely distinguish the type)
- allow for editing and re-editing
- style the summary
- fix linting errors

### chore: test create application
- moved the useCreateOrEditApplication hook to own file to reuse for the admin campaign edit
- added campaign end date handling (if the user selects a date for the end it pops up an input and fills in the preselected by user date or today if none)
- removed some unused props from the cam app details
- added msw v1 https://v1.mswjs.io/docs/getting-started/mocks/rest-api to mock the server responses for testing (Mock Service Worker)
- expanded the test setup to provide Session and QueryClient

### chore: add tests for the createOrEdit application hook

### feat: admin edit
- it uses the same controls/steps as the organizer edit but adds the admin only props on top
- adds a link and copy button for the link where organizer can edit so as to include it if sending a mail to the org (e.g. for more info)
- deletes the admin types - using same as regular
- adds the status dropdown for admin edit
- allows the accept terms checkboxes to be disabled for admin

### feat: actual list of campaign applications
#1842 

## Screenshots:

Before|After
---|---
|https://drive.google.com/file/d/15LWbhsaTiZxSnytUefZ6zHJ-T4w4158e/view?usp=sharing 



## Testing

### Steps to test
- prerequisites - running frontend + api + (docker compose) pg-db keycloak + s3-ninja + file upload config + email intergration
- visit the create/edit campaign for organizers i.e. /campaigns/application and /campaigns/application/:id (i.e. https://localhost:3040/campaigns/application) and create a new campaign 
  - expected 
  - to be able to edit it from the created successful campaign application summary page  
  - to receive an email (working email integration required )
- login as admin and visit admin/campaign-applications 
 - expected - list of applications  
- click on name of an application should navigate to admin edit (e.g. admin/campaign-applications/edit/eb4347a2-c8b4-47f1-83e5-67457b20909c) where everything from organizer edit should be available + admin edit props + the "organizer can edit at" link

### Affected urls
/campaigns/application
/campaigns/application/eb4347a2-c8b4-47f1-83e5-67457b20909c
/admin/campaign-applications
/admin/campaign-applications/edit/eb4347a2-c8b4-47f1-83e5-67457b20909c

### New or updated dependencies:



Dependency name|Previous version|Updated version|Details
---|---|---|---
`msw`|`v1.x.x`|-|DEV mock service worker allows for mocking server responses in tests
